presto-main/src/main/java/com/facebook/presto/security/RefreshingSystemAccessControl.java

```diff
@@ -0,0 +1,275 @@
+ * limitations under the License.
+ */
+
+package com.facebook.presto.security;
+
+import com.facebook.presto.spi.CatalogSchemaName;
+import com.facebook.presto.spi.CatalogSchemaTableName;
+import com.facebook.presto.spi.SchemaTableName;
+import com.facebook.presto.spi.security.Identity;
+import com.facebook.presto.spi.security.Privilege;
+import com.facebook.presto.spi.security.SystemAccessControl;
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import io.airlift.log.Logger;
+
+import java.security.Principal;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+public class RefreshingSystemAccessControl
+        implements SystemAccessControl
+{
+    private static final Logger LOG = Logger.get(RefreshingSystemAccessControl.class);
+
+    static final String REFRESH_PERIOD_SEC = "security.refresh-sec";
+    private static final String DEFAULT_REFRESH_PERIOD_SEC = "300"; // 5 mins
+    static final String REFRESH_ENABLED = "security.refresh-enabled";
+    private static final String DEFAULT_REFRESH = "false";
+    private final ScheduledThreadPoolExecutor service;
+
+    private AtomicReference<SystemAccessControl> delegate = new AtomicReference<>();
+
+    /**
+     * Add optional refreshing to the {@link SystemAccessControl} provided by the factory function. The previous
+     * {@link SystemAccessControl} is provided as an {@link Optional} input - null, when there is no previous element
+     * (as with the first time). If refreshing is not enabled, then the factory function is just used to create the
+     * {@link SystemAccessControl} and returned.
+     *
+     * @param config access control configuration from
+     *         {@link com.facebook.presto.spi.security.SystemAccessControlFactory#create(Map)}
+     * @param factory create the next {@link SystemAccessControl}, optionally informed by the previous
+     * @return a {@link SystemAccessControl} that might be able to refresh itself
+     */
+    static SystemAccessControl optionallyRefresh(Map<String, String> config,
+            Supplier<Optional<SystemAccessControl>> factory)
+    {
+        if (Boolean.valueOf(config.getOrDefault(REFRESH_ENABLED, DEFAULT_REFRESH))) {
+            long period = Long.valueOf(config.getOrDefault(REFRESH_PERIOD_SEC, DEFAULT_REFRESH_PERIOD_SEC));
+            return new RefreshingSystemAccessControl(factory, period);
+        }
+        return factory.get().orElseThrow(() -> new IllegalStateException("Access Control factory did not provide an " +
+                "initial SystemAccessControl."));
+    }
+
+    private RefreshingSystemAccessControl(Supplier<Optional<SystemAccessControl>> factory, long refreshPeriodSec)
```

Comment:
What optional.empty means here? Is it this needed here?

Reply:
I'll add some comments, but empty == no change. Original code had the previous access control passed in to make this easy, but @findepi wasn't as happy about that design